pulp_ansible/app/tasks/collections.py

```diff
@@ -71,22 +71,52 @@ def sync(remote_pk, repository_pk, mirror):
     d_version.create()
 
 
-def import_collection(artifact_pk, repository_pk=None):
+def import_collection(
+    artifact_pk,
+    repository_pk=None,
+    expected_namespace=None,
+    expected_name=None,
+    expected_version=None,
+):
     """
-    Create a Collection from an uploaded artifact.
+    Create a Collection from an uploaded artifact and optionally validate its expected metadata.
+
+    This task provides optional validation of the `namesapce`, `name`, and `version` metadata
+    attributes. If the Artifact fails validation, the Artifact is deleted and the Collection is not
+    created.
 
     Args:
-        artifact_pk (str): The pk or the Artifact to create the Collection from.
+        artifact_pk (str): The pk of the Artifact to create the Collection from.
 
     Keyword Args:
         repository_pk (str): Optional. If specified, a new RepositoryVersion will be created for the
             Repository and any new Collection content associated with it.
+        expected_namespace (str): Optional. The namespace is validated against the namespace
+            specified in the Collection's metadata. If it does not match an AssertionError is
+            raised.
+        expected_name (str): Optional. The name is validated against the name specified in the
+            Collection's metadata. If it does not match an AssertionError is raised.
+        expected_version (str): Optional. The version is validated against the version specified in
+            the Collection's metadata. If it does not match an AssertionError is raised.
     """
     artifact = Artifact.objects.get(pk=artifact_pk)
     log.info("Processing collection from {path}".format(path=artifact.file.path))
     importer_result = process_collection(str(artifact.file.path))
     collection_info = importer_result["metadata"]
 
+    try:
+        if expected_namespace:
+            assert expected_namespace is collection_info["namespace"]
+
+        if expected_name:
+            assert expected_name is collection_info["name"]
+
+        if expected_version:
+            assert expected_version is collection_info["version"]
+    except AssertionError:
+        artifact.delete()
+        raise
```

Comment:
Any reason to raise a generic exception here?

Reply:
I'm open to other suggestions. I think I mostly shyed away from defining a custom Exception object. What would you recommend?

Reply:
I think a generic exception is fine actually. No need to overengineer it.

Reply:
I don't think it's fine. It's not fine at all. 

1. You should **NEVER** use assert operator for business logic.
Assertions are disabled if python is called with `-O` flag. 

2. You should **NEVER** compare strings with `is` operator. `is` operator compares objects identity, not value identity. You cannot rely on internal python optimizations and intern strings for comparison.

Correct code:

```
if (
    expected namespace != collection_info["namespace"]
    or expected_name != collection_info["name"]
    or expected_version != collection_info["version"]
):
    artifact.delete()
    raise ValidationErrorClass("Collection metadata doesn't match expected metadata").

Note: The best would be to put artifact deletion in a context manager, to make sure it's always deleted if exception occurs within the scope.
```

Reply:
Thanks for pointing these things out. I agree and I'll push a fix now.

Reply:
fixed in this push @cutwater lmk what you think.